File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
from collections import defaultdict

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = right = 0
        n = len(s)
        counter = defaultdict(int)
        max_length = 0
        while left < n and right < n:
            if counter[s[right]] == 0:
                counter[s[right]] += 1
                right += 1
                if max_length < right - left:
                    max_length = right - left
            else:
                counter[s[left]] -= 1
                left += 1
        return max_length
            

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.ts
function lengthOfLongestSubstring(s: string): number {
    let left = 0;
    let right = 0;

    const dict = {};
    const n = s.length;
    let result = 0;
    while (left < n && right < n) {
        const c = s[right];
        if (dict[c] === 0 || dict[c] === undefined) {
            dict[c] = 1;
            right += 1;
            result = Math.max(result, right - left);
        } else {
            dict[s[left]] = 0;
            left += 1;
        }
    }
    return result;
};